lib/Sema/PreCheckExpr.cpp

```diff
@@ -761,8 +761,29 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
           UDRE->getNameLoc(), NTD, BaseDC,
           DC->mapTypeIntoContext(NTD->getInterfaceType()));
     } else {
-      BaseExpr = new (Context) DeclRefExpr(Base, UDRE->getNameLoc(),
-                                           /*Implicit=*/true);
+      // If this is an implicit self reference, the `VarDecl` will
+      // currently have a non-optional type. But at this point,
+      // we don't actually know if this is correct (it could be a
+      // weak self capture that hasn't been unwrapped yet).
+      //  - Since we don't know the correct type of self yet,
+      //    we leave this as an UnresolvedDeclRefExpr that is
+      //    populated with the actual type later.
+      bool isClosureImplicitSelfParameter = false;
```

Comment:
Could you please provide an example why this is necessary? Maybe it used to be the case before multi-statement closures but now I don't think we need this, implicit `self` would get a type only when the solver gets to solve for this expression and `guard`, `if let`  effectively shadow global `self` with its local version so it should never be optional if there is guard/let etc.

Reply:
@xedin Thanks for the info! I checked and this is no longer necessary for the original reason -- even without this, implicit `self` decls that haven't been unwrapped have the correct `Optional` type.

Removing this causes issues elsewhere in the implementation, though -- most importantly, [this](https://github.com/apple/swift/pull/40702/files#diff-cb1dfa36aab299d0abafe04242dbebf8a555ad1a20041b47c1299f4687935f41R1619) `var->getParentPatternStmt()` call on the implicit self `VarDecl` is now always null. It sounds like the parent pattern statement is specifically populated during type checking, and changing this `DeclRefExpr` to a `UnresolvedDeclRefExpr` causes the type checker to make another pass on this implicit self decl and populate this key information.

What do you think is the best approach here? If we remove this code that changes the `DeclRefExpr` to a `UnresolvedDeclRefExpr` then we'll need some other way to retrieve (or get the type checker to populate) the `var->getParentPatternStmt()`. Any suggestions on how we could do that?

If this specific block seems ok as-is, I can update the comments to better explain the rationale.

Reply:
There should be a way to populate that during synthesis of the implicit self declaration or at some later point. Returning `UnresolvedDeclRefExpr` from pre-check is not safe because there is nothing in place to guarantee that another pre-check is going to happen and CSGen could be happy to accept UnresolvedDeclRefExpr, create a type variable for it, and infer a type; which then leads to a broken AST and crashes in verifier (in assert builds) or in SILGen (in production compilers).

Reply:
That makes sense, thanks. I updated this to resolve and return the correct `DeclRefExpr` instead of returning an `UnresolvedDeclRefExpr` (https://github.com/apple/swift/pull/40702/commits/9e9f7be0f6a3504ad59b1207664c916f1dd2a5e0). How does that look?

---

The underlying issue here is that implicit self decls always default to being a `ParamDecl` that refers to the closure's `self` param. This was true before, but is no longer true now that `weak self` captures support implicit self. 

For example, here:

```swift
buttonTapped { [weak self] in
  guard let self else { return }
  dismiss() // implicit self
}
```

the implicit self's `VarDecl` should actually come from the guard statement pattern, but without this change its `VarDecl` is the closure's implicit self `ParamDecl`. That explains why `var->getParentPatternStmt()` was always null -- the closure's implicit self `ParamDecl` doesn't have a parent.

Reply:
I think this is incorrect because lookup should be able to determine what is what here - if `self` is used before `if let` or `guard` it should point to parameter declaration from the capture list, if it's used after - to the local variable declaration where parent is going to be the statement that introduced the pattern binding with name `self` without having to re-run resolveDeclRefExpr multiple times in closure context.

Reply:
Is there a way to look up the proper `VarDecl` without using `resolveDeclRefExpr`?

Reply:
What I mean is `lookupUnqualified` is supposed to handle that without any changes.

Reply:
I think it's worth trying to figure out what is going on in lookupUnqualified because implicit "self" in cases after `if let` or `guard` that create a new binding should be pointing to the outer context `self`:

```
func doVoidStuff(_ fn : @escaping () -> ()) {}

public class TestImplicitSelfForWeakSelfCapture {
  static var staticOptional: TestImplicitSelfForWeakSelfCapture? = .init()
  func method() { }

  private init() {
    doVoidStuff { [weak self] in
      guard let self = self else { return }
      method()
    }
  }
}
```

Currently `method()` as "self" that points to `init().self` which I think is incorrect because it needs to point to the "self" in its scope which is the one from `guard let` declaration.

Reply:
@xedin, I updated this in https://github.com/apple/swift/pull/40702/commits/95949cec476d96be7458be8201f1a485387c40e7 to use `ASTScope::lookupUnqualified` (through `ASTScope::lookupSingleLocalDecl`) to look up the proper `VarDecl`. 

It previously wasn't making any lookups at all -- it was just using [`LookupResultEntry::getBaseDecl()`](https://github.com/apple/swift/blob/005c9f34031500404e498c630d803cc8e9e3433e/lib/AST/NameLookup.cpp#L55), which returns the context's `ParamDecl *getImplicitSelfDecl()`. So that explains why it was incorrect before.

Reply:
Can we make it so the lookup itself has local declaration as a base (by changing `getBaseDecl()` maybe)?

Reply:
What do you mean by "the lookup itself has local declaration as a base"? Are you referring to one of the params to the `ASTScope::lookupSingleLocalDecl` call?

Reply:
Sorry, what other lookup are you referring to? `LookupResultEntry::getBaseDecl()` doesn't perform a lookup, and the `LookupResultEntry` itself isn't from a lookup for the implicit self decl (it's a lookup for the explicit member access call that's using the implicit self decl as its base).

Reply:
We could move the `ASTScope::lookupSingleLocalDecl` lookup call into `LookupResultEntry:getBaseDecl()` if you think that's a better way to organize this.

Reply:
>  `LookupResultEntry::getBaseDecl()` should know that it's possible to shadow self with local self declaration and produce results because on whether the UDRE that requires implicit self was located 

This isn't currently possible with the information available in `LookupResultEntry`. We'd either have to pass in information about `resolveDeclRefExpr`'s `UnresolvedDeclRefExpr` and `DeclContext` (which isn't available in all call sites to `LookupResultEntry::getBaseDecl()`), or restructure the API of `LookupResultEntry` (which has a fair number of call sites  that would need to be updated).

Is it possible to leave logic in `resolveDeclRefExpr`, where we already have the context necessary to perform this lookup?

Reply:
From my previous comments you might have gauged that I don't think it's the right place :) If we want to fix it then we should fix it in a robust way even if it requires changes to lookup API, it's not the job of `resolveDeclRef` to modify lookup results.

Reply:
@xedin, that makes sense, thanks. I moved this lookup from `resolveDeclRef` into the implementation of `ASTScope::lookupUnqualified` in https://github.com/apple/swift/pull/40702/commits/03322bf716d94ca2590a915bd2efdd93646808a6. Now `LookupResultEntry::getBaseDecl()` returns the correct decl. How does that look?